### ISSUE TITLE
invalid memory address or nil pointer dereference at s3store.go:928

### ISSUE THREAD

**Issue by Foxcool** [tus/tusd#464]
**Describe the bug**
I am trying to pair this solution with seaweedfs s3.
Probably the problem manifests itself due to an unexpected storage response to the request store.Service.ListPartsWithContext.

		if listPtr.IsTruncated != nil && *listPtr.IsTruncated {
			partMarker = *listPtr.NextPartNumberMarker
		} else {
			break
The condition is fulfilled because IsTruncated true. But NextPartNumber is NIL.

**To Reproduce**
1. Start seaweedfs service with S3 filer.
2. Create service with TUSD s3store
3. Try to upload file

**Stacktrace**
```
downloader    | [tusd] 2021/03/16 13:57:35 event="RequestIncoming" method="PATCH" path="2a0aacb932acdec7b2806319cb7bdd95+d0b86628-b416-4a50-964e-2946243bc40c" requestId="" 
downloader    | 2021/03/16 13:57:35 http: panic serving 172.19.0.1:59978: runtime error: invalid memory address or nil pointer dereference
downloader    | goroutine 48 [running]:
downloader    | net/http.(*conn).serve.func1(0xc0001c9180)
downloader    |         /usr/local/go/src/net/http/server.go:1800 +0x149
downloader    | panic(0xd8d120, 0x140ddb0)
downloader    |         /usr/local/go/src/runtime/panic.go:975 +0x41b
downloader    | gitlab.unim.internal/platform/downloader/vendor/github.com/tus/tusd/pkg/s3store.S3Store.listAllParts(0xc00002bdf0, 0x4, 0x0, 0x0, 0x0, 0x0, 0xfe2400, 0xc000010008, 0x140000000, 0x500000, ...)
downloader    |         /go/src/gitlab.unim.internal/platform/downloader/vendor/github.com/tus/tusd/pkg/s3store/s3store.go:928 +0x5f8
downloader    | gitlab.unim.internal/platform/downloader/vendor/github.com/tus/tusd/pkg/s3store.s3Upload.fetchInfo(0xc001149987, 0x45, 0xc001199c20, 0x0, 0xfdf400, 0xc00002a028, 0xfdf400, 0xc00002a028, 0xc001149987, 0x45, ...)
downloader    |         /go/src/gitlab.unim.internal/platform/downloader/vendor/github.com/tus/tusd/pkg/s3store/s3store.go:572 +0x684
downloader    | gitlab.unim.internal/platform/downloader/vendor/github.com/tus/tusd/pkg/s3store.(*s3Upload).GetInfo(0xc001151d00, 0xfdf400, 0xc00002a028, 0xc001149987, 0x45, 0xfdf600, 0xc001151d00, 0x0, 0x0, 0x460b53, ...)
downloader    |         /go/src/gitlab.unim.internal/platform/downloader/vendor/github.com/tus/tusd/pkg/s3store/s3store.go:540 +0x192
downloader    | gitlab.unim.internal/platform/downloader/vendor/github.com/tus/tusd/pkg/handler.(*UnroutedHandler).PatchFile(0xc00007e0d0, 0xfde3c0, 0xc00119c620, 0xc001aa9600)
downloader    |         /go/src/gitlab.unim.internal/platform/downloader/vendor/github.com/tus/tusd/pkg/handler/unrouted_handler.go:521 +0x7b5
downloader    | net/http.HandlerFunc.ServeHTTP(0xc000020510, 0xfde3c0, 0xc00119c620, 0xc001aa9600)
downloader    |         /usr/local/go/src/net/http/server.go:2041 +0x44
downloader    | gitlab.unim.internal/platform/downloader/vendor/github.com/bmizerany/pat.(*PatternServeMux).ServeHTTP(0xc0001d4120, 0xfde3c0, 0xc00119c620, 0xc001aa9600)
downloader    |         /go/src/gitlab.unim.internal/platform/downloader/vendor/github.com/bmizerany/pat/mux.go:117 +0x313
downloader    | gitlab.unim.internal/platform/downloader/vendor/github.com/tus/tusd/pkg/handler.(*UnroutedHandler).Middleware.func1(0xfde3c0, 0xc00119c620, 0xc001aa9600)
downloader    |         /go/src/gitlab.unim.internal/platform/downloader/vendor/github.com/tus/tusd/pkg/handler/unrouted_handler.go:270 +0xa62
downloader    | net/http.HandlerFunc.ServeHTTP(0xc0001d4140, 0xfde3c0, 0xc00119c620, 0xc001aa9600)
downloader    |         /usr/local/go/src/net/http/server.go:2041 +0x44
downloader    | net/http.StripPrefix.func1(0xfde3c0, 0xc00119c620, 0xc001aa9500)
downloader    |         /usr/local/go/src/net/http/server.go:2080 +0x213
downloader    | net/http.HandlerFunc.ServeHTTP(0xc000105aa0, 0xfde3c0, 0xc00119c620, 0xc001aa9500)
downloader    |         /usr/local/go/src/net/http/server.go:2041 +0x44
downloader    | net/http.(*ServeMux).ServeHTTP(0x1466cc0, 0xfde3c0, 0xc00119c620, 0xc001aa9500)
downloader    |         /usr/local/go/src/net/http/server.go:2416 +0x1a6
downloader    | net/http.serverHandler.ServeHTTP(0xc0011662a0, 0xfde3c0, 0xc00119c620, 0xc001aa9500)
downloader    |         /usr/local/go/src/net/http/server.go:2836 +0x20f
downloader    | net/http.(*conn).serve(0xc0001c9180, 0xfdf3c0, 0xc001a97f80)
downloader    |         /usr/local/go/src/net/http/server.go:1924 +0x173a
downloader    | created by net/http.(*Server).Serve
downloader    |         /usr/local/go/src/net/http/server.go:2962 +0x931
```

**Setup details**
Please provide following details, if applicable to your situation:
- Operating System: Linux, docker
- Used tusd version: github.com/tus/tusd v1.5.1
- Used tusd data storage: s3stor, SeaWeedFS
- Used tus client library: https://github.com/fentas/tus-shell-client

**Comment by Foxcool** [comment]
Debugger screen


![Screenshot_20210316_173619](https://user-images.githubusercontent.com/323923/111327098-246c2300-867e-11eb-8031-e61d823a5335.png)

**Comment by Acconut** [comment]
Thank you for reporting, but this does not look like a bug inside tusd but rather on seaweedfs' side. From the AWS S3 documentation it sounds like NextPartNumberMarker should be set if IsTruncated is true (https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListParts.html#AmazonS3-ListParts-response-NextPartNumberMarker):

>  NextPartNumberMarker
When a list is truncated, this element specifies the last part in the list, as well as the value to use for the part-number-marker request parameter in a subsequent request. 

Seaweedfs has probably not implemented this entirely correct and is therefore not fully S3-compatible (which is a huge undertaking given S3's complexity). This is better fixed at the source of the bug (aka seaweedfs) instead of the place where the symptoms arise (aka tusd).

All in all, the response in your debugger screenshot looks weird. IsTruncated is true, but the Parts object is nil. Maybe their API returns wrong values if a multipart upload does not contain any parts yet.

I recommend you to report this bug/behavior to seaweedfs and hopefully they can resolve this.

**Comment by Foxcool** [comment]
https://github.com/chrislusf/seaweedfs/issues/1914

Okay. I created a ticket there.
I also made a pull request with the addition of a simple check for the emptiness of this field. With it, the download is successful. I don’t know if the lead-additional check, even if it contradicts the description of the protocol, is something redundant. It's up to you to decide.

Thank you very much for your prompt reply!
